### ISSUE TITLE
Ingest S3 Data Structure

### ISSUE THREAD

**Issue by ian-noaa** [NOAA-GSL/unified-graphics#126]
## Problem

There are a number of variations of the RRFS and 3D-RTMA models. We need to be able to identify several characteristics of the diagnostic data that we are ingesting for Unified Graphics:

* Model (RRFS, RTMA, HRRR)
* System (Jet, WCOSS, Hera, etc…)
* Background (only relevant to the 3D-RTMA - could be HRRR or RRFS)
* Domain (North America, CONUS, Alaska, Puerto Rico, Guam, American Samoa, Caribbean, etc…)
* Frequency (Realtime vs Retro)

These fields will need to be stored in our data structures so that modelers can select the diagnostic data for the particular model run in which they are interested.

## Solution

Given the above identifiers, we think a useful S3 key for incoming data would look like:

```
<model>/<system>/<domain>/<frequency>/<filename>_<background>.nc4
```

It’s important to reiterate, this is the definition for an S3 key prefix for data uploaded from models for ingest into Unified Graphics, this structure does not dictate how we will store the diagnostic data for use by the Unified Graphics application.

## No Gos

We will focus on the below primary/baseline model versions. We won't support arbitrary model runs yet.
1. RRFS_A (from WCOSS, currently CONUS, will switch to North American version soon)
2. RTMA Versions (note these are all 3D versions, only version that is not 3D is the operational version that we want to replace):
    1. RTMA_A (from WCOSS, uses RRFS_A background data so also currently CONUS will switch to North American version soon)
    2. RTMA_C (from cloud, uses RRFS_A background data so also currently CONUS will switch to North American version soon)
    3. RTMA_HRRR_Baseline, also call legacy RTMA (from WCOSS)

At this point in time, we won't require additional metadata in the netcdf files. However, this would be useful for capturing secondary identifiers like model developer, git SHA, git branch, "time to live", and other notes about what causes this model to differ.

Likewise, we also won't support exposing GSI namelists. These can provide important information about what parameters have been modified but don't capture the whole picture and will add additional parsing. 

## Rabbit Holes

_Describe any solutions to problems that pose a risk to completing this project on time._

**Comment by esheehan-gsl** [comment]
@ian-noaa I have a proposal for organizing the NetCDF files for ingest. Let me know what you think.

I was thinking we could encode the model, system, and domain as directories. Background, I’d suggest including in the filename, since it’s only relevant to RTMA runs (we could also ask for that in the NetCDF attributes eventually).

For frequency, I could go either way: include it in the file name (or NetCDF attributes), or as another directory level.

So the directory structure would look like this:

```
<model>/<system>/<domain>/<filename>_<background>.nc4
```

**Comment by ian-noaa** [comment]
Thanks, @esheehan-gsl. That was similar to what I had in mind and I think that sounds like a good place to start. I need to do a bit more reading up on how S3 keys/"directory structures" work and what sort of tools AWS provides for filtering them. It sounds like AWS provides an [API that lets you filter object keys by the key prefix](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.list_objects_v2). E.g. you could ask for all the keys in the bucket that start with `RRFS/WCOSS/CONUS/`. However, it may be difficult to just get the "directory structure".) I could see us starting with this and iterating on it as we learn more about our data usage.

That's a good point that netcdf attributes would be a good fit for model-specific metadata. It would certainly simplify our structure if we could limit the keys to properties that were common between models. 

I could see frequency being in either location. However, if it's likely that we'd want to persist retros longer than realtime output we'd probably want to put it in the key - that way we could use that information in a lifecycle policy.

And a point of clarification (I've had too many things on my mind... my apologies) - is this structure something we're hoping our data source buckets will implement, for our "internal" S3 bucket, or for both?

**Comment by esheehan-gsl** [comment]
Do we need to get just the directory structure? If this is just how we store the NetCDF files for ingest — the ones we convert to Zarr — then could we just get a list of new files along with their keys and parse the whole key?

```py
model, system, domain, freq, filename = key.split("/")
```

We could either poll the S3 buckets if we have to, and ask for files newer than our last poll, or we could listen for new file events and just run on each file as it comes in. Are either of these an option? I seem to remember that there is an event system for starting Lambdas in response to changes to S3 buckets.

On our end, once we’re storing them in Zarr, we can create groups in the Zarr array that reflect this organization: model, system, domain, frequency. Zarr, if I recall correctly, will give us an interface for listing groups in the array, which will be useful for populating selects in the UI for people to choose model, system, _etc._

Does that make sense?

Also, your point about retention sold me on putting realtime/retro in the key, not the file name. So let’s do that. It should make it easier to manage retention.

**Comment by ian-noaa** [comment]
That's a great point. I was thinking about populating the UI during my first comment so was mostly thinking about how we'd go about that instead of how things would work for ingesting the data and transforming it to Zarr.

For polling v eventing - I think polling the source buckets would be complicated since I don't _think_ AWS has a simple mechanism to denote a "new" object since the last time the bucket was listed. We'd most likely have to track previously processed objects ourselves. However, since this isn't being used to populate the UI, we can use SQS events from the input buckets to trigger our processing. The application would then have the latest files/objects along with the metadata contained in the key as you proposed with:

```python
model, system, domain, freq, filename = key.split("/")
```

On the API/UI side - if we can query the Zarr array to get the available groups out that would be quite helpful. It seems like having a "known" key structure would be useful as querying the bucket for all the keys would probably be slow. (Or maybe we store this info in some other store?) I may also be overthinking this. 😅

**Comment by esheehan-gsl** [comment]
I think the Zarr will do the trick for us in the application. But if it doesn’t, I think you’re on the right track when it comes to building a separate store that has metadata regarding which model runs are available.

**Comment by esheehan-gsl** [comment]
@ian-noaa I made some changes to the original proposal to clarify that we’re talking about the key prefix for uploaded NetCDF files for ingest and not for how we store the data for our own use after ingest.

I also added the frequency (realtime/retro) to the prefix. Even if everything goes in realtime for now, it seems like a good idea to anticipate the need to support retros without have to redo/backfill a bunch of existing data.

If this looks good to you, I think we can move this out of shaping.
